freegs/reconstruction.py

```diff
@@ -0,0 +1,803 @@
+        row_num = 0
+        for i in range(1, m - 1):
+            r_h = r_FE[i]
+            for j in range(1, n - 1):
+                z_h = z_FE[j]
+
+                for h in range(T.shape[0]):
+                    if (abs(R[h] - r_h) / dR) < 1:
+                        if (abs(Z[h] - z_h) / dZ < 1):
+                            T[h, row_num] = (1 - abs(R[h] - r_h) / dR) * (
+                                        1 - abs(Z[h] - z_h) / dZ)
+                row_num += 1
+
+        return T
+
+    # Indexing function
+    def get_index(self, Vector, indexing):
+
+        # Finding the corresponding index for the string
+        if indexing == 'plasma':
+            if self.use_VerticalControl and self.it>1:
+                return Vector[:(self.pprime_order+self.ffprime_order+1)]
+            else:
+                return Vector[:(self.pprime_order+self.ffprime_order)]
+        if indexing == 'coils':
+            if self.use_VesselCurrents:
+                return Vector[-(self.tokamak.n_coils+self.tokamak.n_basis):-self.tokamak.n_basis]
+            else:
+                return Vector[-self.tokamak.n_coils:]
+
+        if indexing == 'vessel':
+            return Vector[-self.tokamak.n_basis:]
+
+    # Reconstruction Algorithm
+    def reconstruct(self):
+        """
+        Parameters
+        ----------
+        tokamak - tokamak object
+        M - measurements
+        sigma - measurement uncertainties
+        pprime_order - number of polynomial coefficients for pprime model
+        ffprime_order - number of polynomial coefficients for ffprime model
+        tolerance - value beneath which the chi squared value will stop the iteration
+        Gplasma - Plasma Response Greens matrix
+        Gcoil - Coil Greens matrix
+        Gvessel - Eigenbasis response Greens matrix
+        show - option to display psi through the iterations
+        pause - time to wait between iterations
+        use_VesselCurrents - Used when tokamak has vessel filaments. Allows calculation of induced currents in tokamak wall
+        use_VerticalControl - option for use of the vertical control found in ferron's paper
+        use_CurrentInitialisation - option for using finite element method to optimise initial guess for jtor
+        """
+
+        # start iterative loop
+        self.it = 0  # iteration number
+        convergence = False
+        while not convergence:
+
+            # Initialisation
+            if self.it == 0:
+                jtor_2d = None
+
+                if self.show:
+                    import matplotlib.pyplot as plt
+                    from .plotting import plotEquilibrium
+
+                    if self.pause > 0.0:
+                        # No axis specified, so create a new figure
+                        fig, axs = plt.subplots(1,2, sharex=True, sharey=True)
+
+                # Performs plasma current density initialisation
+                if self.use_CurrentInitialisation:
```

Comment:
what if its false?

Reply:
then no initial guess for current density is made